VCS: Port Godot 3.5's VCS features to GDExtension

twaritwaikar opened the pull request:
Pull the newer VCS features and an improved diff viewer from 3.5 to 4

Watch this demo for this intended usage - https://youtu.be/blVtgs4g_GU

I was able to port ALL of the features as they exist in 3.5. The next task is to pull out the UI into the plugin as per https://github.com/godotengine/godot-proposals/issues/4369, but in case that turns out to be too difficult, getting this PR merged will ensure the VCS integration works as expected.

### Issues

I tried some fixes to clean the implementation up but there are still some issues that don't seem to be related to the changes in this PR but they can be put into newer GH issues:

* If a script has been opened in the code editor, and you try to write a commit message, currently, the code editor will _steal_ focus from the commit message textbox on every keystroke, making it impossible to write a proper commit message. I have to switch to the 3D tab or the 2D tab to write a commit message.

![image](https://user-images.githubusercontent.com/31801364/182036667-978f04e8-8616-43a2-b549-3be681f04ab5.png)
![image](https://user-images.githubusercontent.com/31801364/182036717-ba874f20-3c36-44a1-b522-4e0bd685703b.png)

* `RichTextLabel::append_text()` is not making the text `[center]`'d properly. The text surrounded by `@@` should have been centered.
* The `push_indent` within `RichTextLabel`'s table cells is not working as expected in the split view. The line numbers are overlapping with the content. This doesn't appear to happen in the unified view.

twaritwaikar committed:
VCS: Port Godot 3.5's VCS features to GDExtension

twaritwaikar committed:
VCS: Load a GDExtension instead of Script + UI port from 3.5

twaritwaikar committed:
VCS: Fix empty metadata items being moved across stages

ChronicallySerious commented:
Currently godot-cpp is crashing after my recent rebase on master. Reported https://github.com/godotengine/godot-cpp/issues/800

twaritwaikar committed:
VCS: Replace usage of varray() by using callable_mp::bind()

ChronicallySerious commented:
I have added a demo of the changes in my PR description

twaritwaikar committed:
VCS: Improve VCS UI/UX by QoL changes

The editor will now use the project path i.e. the place where the root of
the repo is supposed to be according to the user. This project path is
also sent into the plugin and so out-of-directory asset folders can also be
maintained this way.

akien-mga commented on editor/editor_vcs_interface.cpp:
```diff
@@ -30,132 +30,371 @@
 
+#include "editor_node.h"
+
+#define UNIMPLEMENTED() ERR_PRINT(TTR("Unimplemented virtual function in EditorVCSInterface based node: ") + __func__)
```
We normally don't translate console errors, so this shouldn't use `TTR`.
For errors displayed in the editor (e.g. as you do below with `EditorNode::show_warning()`) you should use `TTR` indeed.

akien-mga commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -30,47 +30,81 @@
 
-#include "core/object/script_language.h"
+#include "core/config/project_settings.h"
+#include "core/core_bind.h"
```
I couldn't spot why it's included in the file but in general `core/core_bind.h` shouldn't be used outside core bindings. It's only meant to wrap some core functionality in a way that's convenient for GDScript / other scripting languages. But for C++ you should use the underlying core functionality directly.

akien-mga commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -141,8 +143,8 @@ void VersionControlEditorPlugin::_initialize_vcs() {
 
-		ProjectSettings::get_singleton()->set("editor/version_control/autoload_on_startup", true);
-		ProjectSettings::get_singleton()->set("editor/version_control/plugin_name", selected_plugin);
+		ProjectSettings::get_singleton()->set(SNAME("editor/version_control/autoload_on_startup"), true);
+		ProjectSettings::get_singleton()->set(SNAME("editor/version_control/plugin_name"), selected_plugin);
```
This doesn't need to use `SNAME()`. `SNAME()` allocates static memory and should only be used in critical code paths which are used a lot. I assume this one only runs once so it's fine to let it convert the strings to StringName on the fly.

akien-mga commented on editor/plugins/version_control_editor_plugin.cpp:
```diff
@@ -169,34 +171,27 @@ void VersionControlEditorPlugin::_set_credentials() {
-
-	EditorVCSInterface::set_singleton(vcs_interface);
+	EditorVCSInterface *vcs_plugin = Object::cast_to<EditorVCSInterface>(extension_instance);
+	ERR_FAIL_NULL_V_MSG(vcs_plugin, false, vformat(TTR("Could not cast VCS extension instance to %s."), EditorVCSInterface::get_class_static()));
```
I won't comment on all of them but you should remove all `TTR()` in `ERR_FAIL.*` macros.

As you can see before this PR there's almost no code translating `ERR_FAIL` errors:
```
$ rg ERR_FAIL.*TTR
editor/editor_vcs_interface.cpp
173:                    ERR_FAIL_MSG(TTR("Couldn't create .gitignore in project path."));
180:                    ERR_FAIL_MSG(TTR("Couldn't create .gitattributes in project path."));

editor/import/resource_importer_imagefont.cpp
115:    ERR_FAIL_COND_V_MSG(err != OK, ERR_FILE_CANT_READ, TTR("Can't load font texture:") + " \"" + p_source_file + "\".");
```

akien-mga reviewed (approved):
This is a pretty big PR so not trivial to review, but overall it seems well made, and pretty self-contained so there should be no major problem to merge it.

ChronicallySerious commented:
Also, @groud had reviewed some of the code that touches things outside of the VCS on [chat](https://chat.godotengine.org/channel/editor?msg=K9tXbkXwXN6yJ692n). There seems to be another bug where the CodeEditor is not updating the script files if they get changed from outside the editor (at least on Windows). I will open a separate bug to track this

ChronicallySerious commented:
Once that is solved, we may not need to call `ScriptEditor::reload_scripts()` through this code. There needs to be additional testing on the behavior in that case.

akien-mga commented:
Might need a rebase after #64377 I think.

ChronicallySerious commented:
@akien-mga Rebased, and replaced `update()` with `queue_redraw()`

akien-mga commented:
Thanks!
